iothub/service/src/Query/QueryResponse.cs

```diff
@@ -1,45 +1,131 @@
+        /// while (await queriedTwins.MoveNextAsync())
+        /// {
+        ///     Twin queriedTwin = queriedTwins.Current;
+        ///     Console.WriteLine(queriedTwin);
+        /// }
+        /// </c>
+        /// </example>
+        /// <remarks>
+        /// Like with a more typical implementation of IEnumerator, this function should be called once before checking
+        /// <see cref="Current"/>.
+        ///
+        /// This function is async because it may make a service request to fetch the next page of results if the current page
+        /// of results has been advanced through already.
+        /// </remarks>
+        public async Task<bool> MoveNextAsync(QueryOptions queryOptions = null, CancellationToken cancellationToken = default)
         {
-            return _queryResults.GetEnumerator();
+            if (_items.MoveNext())
+            {
+                // Current page of results still had an item to return to the user.
+                Current = _items.Current;
+                return true;
+            }
+
+            if (ContinuationToken == null && queryOptions?.ContinuationToken == null)
+            {
+                // The query has no more pages of results to return and the last page has been
+                // completely exhausted.
+                return false;
+            }
+
+            // User's can pass in a continuation token themselves, but the default behavior
+            // is to use the continuation token saved by this class when it last retrieved a page.
+            var queryOptionsClone = new QueryOptions()
+            {
+                ContinuationToken = queryOptions?.ContinuationToken ?? ContinuationToken,
+                PageSize = queryOptions?.PageSize ?? _defaultPageSize,
+            };
+
+            //TODO document thrown exceptions
+            if (!string.IsNullOrEmpty(_originalQuery))
+            {
+                QueryResponse<T> response = await _client.CreateAsync<T>(_originalQuery, queryOptionsClone, cancellationToken);
+                CurrentPage = response.CurrentPage;
+                Current = CurrentPage.GetEnumerator().Current;
+                ContinuationToken = response.ContinuationToken;
+            }
+            else
+            {
+                // Job type and job status may still be null here, but that's okay
+                QueryResponse<ScheduledJob> response = await _client.CreateAsync(_jobType, _jobStatus, queryOptionsClone, cancellationToken);
+                CurrentPage = (IEnumerable<T>)response.CurrentPage;
+                Current = CurrentPage.GetEnumerator().Current;
+                ContinuationToken = response.ContinuationToken;
+            }
+
+            return true;
         }
+
+        /// <summary>
+        /// Get the current query result. Can be called multiple times without advancing the query.
```

Comment:
Consider rewording. This looks like the item is the entire query result, not just the current enumerated item in the current page.